tests/reset_test.go

```diff
@@ -0,0 +1,92 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2019 Red Hat, Inc.
+ *
+ */
+
+package tests_test
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/libvmops"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	expect "github.com/google/goexpect"
+
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/tests/console"
+	"kubevirt.io/kubevirt/tests/framework/kubevirt"
+	"kubevirt.io/kubevirt/tests/framework/matcher"
+	"kubevirt.io/kubevirt/tests/libvmifact"
+	"kubevirt.io/kubevirt/tests/testsuite"
+)
+
+func waitForVMIReset(vmi *v1.VirtualMachineInstance) error {
+	By(fmt.Sprintf("Waiting for vmi %s reset", vmi.Name))
+	if vmi.Namespace == "" {
+		vmi.Namespace = testsuite.GetTestNamespace(vmi)
+	}
+	return console.SafeExpectBatch(vmi, []expect.Batcher{
+		&expect.BSnd{S: "\n"},
+		&expect.BExp{R: ".*Detected virtualization kvm.*"},
+	}, 300)
+}
+
+var _ = Describe("[level:component][sig-compute]Reset", decorators.SigCompute, func() {
+	const vmiLaunchTimeout = 360
+
+	var virtClient kubecli.KubevirtClient
+
+	BeforeEach(func() {
+		virtClient = kubevirt.Client()
+	})
+
+	It("reset vmi with should succeed", func() {
+		vmi := libvmops.RunVMIAndExpectLaunch(libvmifact.NewFedora(), vmiLaunchTimeout)
+
+		Eventually(matcher.ThisVMI(vmi), 12*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
+
+		errChan := make(chan error)
+		go func() {
+			time.Sleep(5)
```

Comment:
IIUC we sleep to ensure that the `waitForVMIReset(vmi)` captures the reset.
Can't we simply run it before sending the reset request?

Reply:
Or even better, can't we rely on the virt-handler event?
```
	It("should succeed", func() {
		vmi := libvmops.RunVMIAndExpectLaunch(libvmifact.NewFedora(), vmiLaunchTimeout)

		Eventually(matcher.ThisVMI(vmi), 12*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))

		Expect(virtClient.VirtualMachineInstance(testsuite.GetTestNamespace(vmi)).Reset(context.Background(), vmi.Name)).To(Succeed())
		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()
		watcher.New(vmi).Timeout(60*time.Second).SinceWatchedObjectResourceVersion().WaitFor(ctx, watcher.NormalEvent, "Reset")
	})
```

Reply:
>Or even better, can't we rely on the virt-handler event?

The way this is written now is verifying the `reset` from the perspective of the user. We're verifying that the guest os actually cycles, which is important. A virt-handler event doesn't guarantee this.


> IIUC we sleep to ensure that the waitForVMIReset(vmi) captures the reset.
Can't we simply run it before sending the reset request?

"simply run it"... what's "it"? The wait for reset function? That's a blocking function that's waiting for the reset to be detected. 

I dont' think it matters if we put the "wait for reset" logic in a goroutine or the "reset" action in a goroutine. As long as a goroutine is involved, there will need to be a sleep or some other kind of coordination to ensure we catch the console output.